cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/FieldLineageAdmin.java

```diff
@@ -207,6 +253,30 @@ FieldLineageDetails getOperationDetails(Constants.FieldLineage.Direction directi
     return new FieldLineageDetails(incoming, outgoing);
   }
 
+  private Set<String> getFieldsWithNoFieldLineage(EndPoint dataset,
+                                                  Set<String> lineageFields) throws IOException {
+    // get the system properties of this dataset
+    Map<String, String> properties = metadataAdmin.getProperties(MetadataScope.SYSTEM,
+                                                                 MetadataEntity.ofDataset(dataset.getNamespace(),
+                                                                                          dataset.getName()));
+    // the system metadata contains the schema of the dataset which is written by the DatasetSystemMetadataWriter
+    if (properties.containsKey(MetadataConstants.SCHEMA_KEY)) {
+      String schema = properties.get(MetadataConstants.SCHEMA_KEY);
+      Schema sc = Schema.parseJson(schema);
+      if (sc.getFields() != null) {
+        Set<String> schemaFields = sc.getFields().stream().map(Schema.Field::getName).collect(Collectors.toSet());
+        // Sets.difference will return all the fields which are present in schemaFields and not present in lineage
+        // fields. If there are common fields then they will not be present in the difference and will be treated
+        // as lineage fields containing lineage information.
+        return Sets.difference(schemaFields, lineageFields).immutableCopy();
```

Comment:
You don't need to do a collect and then difference. You can do it with one stream.

```
sc.getFields().stream()
  .map(Schema.Field::getName)
  .filter(name -> !lineageFields.contains(name))
  .collect(Collectors.toSet());
```